// meshapi-node-sdk/dist/index.cjs
'use strict';

// src/errors.ts
var MeshAPIApiError = class _MeshAPIApiError extends Error {
  /** HTTP status code (e.g. 401, 429, 500). 0 means the response body could not be parsed. */
  status;
  /**
   * Machine-readable error code slug from the API.
   *
   * Known values: "unauthorized", "forbidden", "not_found", "model_not_found",
   * "validation_error", "unprocessable_entity", "rate_limit_exceeded",
   * "spend_limit_exceeded", "upstream_error", "gateway_timeout",
   * "internal_error", "parse_error"
   */
  errorCode;
  /** The `request_id` from the response envelope (format: `req_<ULID>`). */
  requestId;
  /** Validation error details array (present when errorCode is "validation_error"). */
  details;
  /** Upstream provider error info when the gateway itself is reporting a proxy failure. */
  providerError;
  /**
   * Number of seconds to wait before retrying.
   * Present when errorCode is "rate_limit_exceeded".
   */
  retryAfterSeconds;
  constructor(status, envelope) {
    super(envelope.error.message);
    this.name = "MeshAPIApiError";
    this.status = status;
    this.errorCode = envelope.error.code;
    this.requestId = envelope.request_id;
    this.details = envelope.error.details ?? [];
    this.providerError = envelope.error.provider_error;
    this.retryAfterSeconds = envelope.error.retry_after_seconds;
    const captureStackTrace = Error["captureStackTrace"];
    if (typeof captureStackTrace === "function") {
      captureStackTrace(
        this,
        _MeshAPIApiError
      );
    }
  }
  /**
   * Attempt to build a MeshAPIApiError from an HTTP response.
   * Falls back to a synthetic "parse_error" if the body is not valid JSON
   * or does not match the expected envelope shape.
   */
  static async fromResponse(response) {
    const contentType = response.headers.get("content-type") ?? "";
    if (contentType.includes("application/json")) {
      try {
        const body = await response.json();
        if (body.error && typeof body.error.code === "string") {
          return new _MeshAPIApiError(response.status, body);
        }
      } catch {
      }
    }
    let rawText = "";
    try {
      rawText = await response.text();
    } catch {
    }
    const syntheticEnvelope = {
      error: {
        code: "parse_error",
        message: rawText.slice(0, 500) || `HTTP ${response.status}`
      },
      request_id: response.headers.get("x-request-id") ?? ""
    };
    return new _MeshAPIApiError(response.status, syntheticEnvelope);
  }
};

// src/http.ts
var RETRY_STATUS_CODES = /* @__PURE__ */ new Set([429, 502, 503, 504]);
var BACKOFF_BASE_MS = 500;
var BACKOFF_MAX_MS = 3e4;
var SDK_VERSION_HEADER = "X-MeshAPI-SDK";
var SDK_VERSION_VALUE = "node/0.1.0";
var HttpClient = class {
  baseUrl;
  token;
  defaultTimeoutMs;
  defaultSignal;
  fetchImpl;
  maxRetries;
  constructor(config) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.token = config.token;
    this.defaultTimeoutMs = config.timeoutMs ?? 6e4;
    this.defaultSignal = config.signal;
    this.fetchImpl = config.fetch ?? globalThis.fetch.bind(globalThis);
    this.maxRetries = config.maxRetries ?? 3;
  }
  async get(path, opts) {
    return this.request("GET", path, void 0, opts);
  }
  async post(path, body, opts) {
    return this.request("POST", path, body, opts);
  }
  async patch(path, body, opts) {
    return this.request("PATCH", path, body, opts);
  }
  async delete(path, opts) {
    await this.request("DELETE", path, void 0, opts);
  }
  async getBytes(path, opts) {
    const response = await this.requestRaw("GET", path, void 0, opts);
    return new Uint8Array(await response.arrayBuffer());
  }
  /**
   * Initiate a streaming request and return the raw Response.
   * The timeout signal here covers only the initial connection (TTFB);
   * once the stream starts, it is the caller's responsibility to cancel via AbortSignal.
   */
  async stream(path, body, opts) {
    const timeoutMs = opts?.timeoutMs ?? this.defaultTimeoutMs;
    const connectSignal = AbortSignal.timeout(timeoutMs);
    const signals = [connectSignal];
    if (opts?.signal) signals.push(opts.signal);
    if (this.defaultSignal) signals.push(this.defaultSignal);
    const signal = signals.length === 1 ? signals[0] ?? connectSignal : AbortSignal.any(signals);
    const response = await this.fetchImpl(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: this.buildHeaders(),
      body: JSON.stringify(body),
      signal
    });
    if (!response.ok) {
      throw await MeshAPIApiError.fromResponse(response);
    }
    return response;
  }
  // ── Private ─────────────────────────────────────────────────────────────────
  buildHeaders() {
    return {
      "Authorization": `Bearer ${this.token}`,
      "Content-Type": "application/json",
      "Accept": "application/json",
      [SDK_VERSION_HEADER]: SDK_VERSION_VALUE
    };
  }
  buildSignal(opts) {
    const timeoutMs = opts?.timeoutMs ?? this.defaultTimeoutMs;
    const timeoutSignal = AbortSignal.timeout(timeoutMs);
    const signals = [timeoutSignal];
    if (opts?.signal) signals.push(opts.signal);
    if (this.defaultSignal) signals.push(this.defaultSignal);
    return signals.length === 1 ? signals[0] ?? timeoutSignal : AbortSignal.any(signals);
  }
  async request(method, path, body, opts) {
    const signal = this.buildSignal(opts);
    const init = {
      method,
      headers: this.buildHeaders(),
      signal
    };
    if (body !== void 0) {
      init.body = JSON.stringify(body);
    }
    let attempt = 0;
    while (true) {
      const response = await this.fetchImpl(`${this.baseUrl}${path}`, init);
      if (response.status === 204) {
        return void 0;
      }
      if (RETRY_STATUS_CODES.has(response.status) && attempt < this.maxRetries) {
        const delay = this.computeDelay(attempt, this.getRetryAfter(response));
        await new Promise((resolve) => setTimeout(resolve, delay));
        attempt++;
        continue;
      }
      if (!response.ok) {
        throw await MeshAPIApiError.fromResponse(response);
      }
      const contentType = response.headers.get("content-type") ?? "";
      if (!contentType.includes("application/json")) {
        return await response.text();
      }
      return response.json();
    }
  }
  async requestRaw(method, path, body, opts) {
    const signal = this.buildSignal(opts);
    const init = {
      method,
      headers: this.buildHeaders(),
      signal
    };
    if (body !== void 0) {
      init.body = JSON.stringify(body);
    }
    let attempt = 0;
    while (true) {
      const response = await this.fetchImpl(`${this.baseUrl}${path}`, init);
      if (RETRY_STATUS_CODES.has(response.status) && attempt < this.maxRetries) {
        const delay = this.computeDelay(attempt, this.getRetryAfter(response));
        await new Promise((resolve) => setTimeout(resolve, delay));
        attempt++;
        continue;
      }
      if (!response.ok) {
        throw await MeshAPIApiError.fromResponse(response);
      }
      return response;
    }
  }
  computeDelay(attempt, retryAfterMs) {
    const baseMs = retryAfterMs ?? BACKOFF_BASE_MS * Math.pow(2, attempt);
    const capped = Math.min(baseMs, BACKOFF_MAX_MS);
    const jitter = capped * (0.8 + Math.random() * 0.4);
    return jitter;
  }
  getRetryAfter(response) {
    const header = response.headers.get("retry-after");
    if (!header) return null;
    const seconds = parseFloat(header);
    return isNaN(seconds) ? null : Math.ceil(seconds) * 1e3;
  }
};
async function* parseSSEStream(response) {
  yield* parseJSONSSEStream(response);
}
async function* parseJSONSSEStream(response) {
  if (!response.body) {
    throw new Error("Response body is null; cannot parse SSE stream.");
  }
  const reader = response.body.getReader();
  const decoder = new TextDecoder("utf-8", { fatal: false });
  let remainder = "";
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        if (remainder.trim()) {
          const chunk = tryParseJSONSSEFrame(remainder);
          if (chunk !== null) yield chunk;
        }
        break;
      }
      remainder += decoder.decode(value, { stream: true });
      const frames = remainder.split("\n\n");
      remainder = frames.pop() ?? "";
      for (const frame of frames) {
        if (!frame.trim()) continue;
        const chunk = tryParseJSONSSEFrame(frame);
        if (chunk !== null) yield chunk;
      }
    }
  } finally {
    reader.releaseLock();
  }
}
function tryParseJSONSSEFrame(frame) {
  const dataLines = [];
  for (const line of frame.split("\n")) {
    if (line.startsWith("data: ")) {
      dataLines.push(line.slice(6).trim());
    }
  }
  if (dataLines.length === 0) return null;
  for (const data of dataLines) {
    if (data === "[DONE]") return null;
    let parsed;
    try {
      parsed = JSON.parse(data);
    } catch {
      continue;
    }
    if (!isRecord(parsed)) continue;
    if ("error" in parsed && isRecord(parsed["error"])) {
      const err = parsed["error"];
      const requestId = typeof parsed["request_id"] === "string" ? parsed["request_id"] : "";
      throw new MeshAPIApiError(0, {
        error: {
          code: typeof err["code"] === "string" ? err["code"] : "upstream_error",
          message: typeof err["message"] === "string" ? err["message"] : "An upstream error occurred in the stream."
        },
        request_id: requestId
      });
    }
    if (typeof parsed["type"] === "string" && parsed["type"].startsWith("response.")) {
      continue;
    }
    return parsed;
  }
  return null;
}
function isRecord(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function makeLazySSEIterable(http, path, params, opts) {
  return {
    [Symbol.asyncIterator]() {
      let iterator = null;
      const init = async () => {
        const response = await http.stream(path, params, opts);
        const gen = parseSSEStream(response);
        return gen[Symbol.asyncIterator]();
      };
      return {
        async next() {
          if (!iterator) {
            iterator = await init();
          }
          return iterator.next();
        },
        async return(value) {
          if (iterator?.return) {
            return iterator.return(value);
          }
          return { done: true, value: void 0 };
        },
        async throw(err) {
          if (iterator?.throw) {
            return iterator.throw(err);
          }
          throw err;
        }
      };
    }
  };
}

// src/resources/chat.ts
var ChatCompletionsResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  create(params, opts) {
    if (params.stream === true) {
      return this.streamCreate(params, opts);
    }
    return this.http.post("/v1/chat/completions", params, opts);
  }
  streamCreate(params, opts) {
    return makeLazySSEIterable(this.http, "/v1/chat/completions", params, opts);
  }
};
var ChatResource = class {
  completions;
  constructor(http) {
    this.completions = new ChatCompletionsResource(http);
  }
};

// src/resources/compare.ts
var CompareResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  create(params, opts) {
    if (params.stream === true) {
      return this.streamCreate(params, opts);
    }
    return this.http.post("/v1/chat/compare", params, opts);
  }
  streamCreate(params, opts) {
    const http = this.http;
    return {
      [Symbol.asyncIterator]() {
        let iterator = null;
        const init = async () => {
          const response = await http.stream("/v1/chat/compare", params, opts);
          const gen = parseJSONSSEStream(response);
          return gen[Symbol.asyncIterator]();
        };
        return {
          async next() {
            if (!iterator) iterator = await init();
            return iterator.next();
          },
          async return(value) {
            if (iterator?.return) return iterator.return(value);
            return { done: true, value: void 0 };
          },
          async throw(err) {
            if (iterator?.throw) return iterator.throw(err);
            throw err;
          }
        };
      }
    };
  }
};

// src/resources/embeddings.ts
var EmbeddingsResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  create(params, opts) {
    return this.http.post("/v1/embeddings", params, opts);
  }
};

// src/resources/files.ts
var FilesResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  upload(params, opts) {
    return this.http.post("/v1/files", params, opts);
  }
  get(fileId, opts) {
    return this.http.get(`/v1/files/${fileId}`, opts);
  }
  delete(fileId, opts) {
    return this.http.delete(`/v1/files/${fileId}`, opts);
  }
  content(fileId, opts) {
    return this.http.getBytes(`/v1/files/${fileId}/content`, opts);
  }
};

// src/resources/batches.ts
var BatchesResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  create(params, opts) {
    return this.http.post("/v1/batches", params, opts);
  }
  list(params = {}, opts) {
    const query = new URLSearchParams();
    if (params.after) query.set("after", params.after);
    if (params.limit !== void 0) query.set("limit", String(params.limit));
    const suffix = query.size ? `?${query.toString()}` : "";
    return this.http.get(`/v1/batches${suffix}`, opts);
  }
  get(batchId, opts) {
    return this.http.get(`/v1/batches/${batchId}`, opts);
  }
  cancel(batchId, opts) {
    return this.http.post(`/v1/batches/${batchId}/cancel`, {}, opts);
  }
};

// src/resources/models.ts
var ModelsResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /**
   * List all available models.
   *
   * @param params - Optional filter: `{ free: true }` for free models only,
   *   `{ free: false }` for paid only, omit for all.
   *
   * Auth: API key (`rsk_...`) or Supabase JWT
   *
   * @example
   * ```ts
   * const models = await client.models.list();
   * const freeOnly = await client.models.list({ free: true });
   * ```
   */
  async list(params, opts) {
    const qs = params?.free !== void 0 ? `?free=${params.free}` : "";
    return this.http.get(`/v1/models${qs}`, opts);
  }
  /**
   * List only free-tier models (zero prompt + completion cost).
   *
   * Auth: API key (`rsk_...`) or Supabase JWT
   *
   * @example
   * ```ts
   * const freeModels = await client.models.free();
   * ```
   */
  async free(opts) {
    return this.http.get("/v1/models/free", opts);
  }
  /**
   * List only paid models.
   *
   * Auth: API key (`rsk_...`) or Supabase JWT
   *
   * @example
   * ```ts
   * const paidModels = await client.models.paid();
   * ```
   */
  async paid(opts) {
    return this.http.get("/v1/models/paid", opts);
  }
};

// src/resources/responses.ts
var ResponsesResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  create(params, opts) {
    if (params.stream === true) {
      return this.streamCreate(params, opts);
    }
    return this.http.post("/v1/responses", params, opts);
  }
  streamCreate(params, opts) {
    return makeLazySSEIterable(this.http, "/v1/responses", params, opts);
  }
};

// src/resources/templates.ts
var TemplatesResource = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /**
   * Create a new prompt template.
   *
   * Auth: Supabase JWT or API key (`rsk_...`)
   * Templates are scoped to the authenticated owner.
   *
   * @example
   * ```ts
   * const template = await client.templates.create({
   *   name: "support-agent",
   *   system: "You are a helpful customer support agent for {{company}}.",
   *   variables: ["company"],
   *   model: "openai/gpt-4o-mini",
   * });
   * console.log(template.id);
   * ```
   */
  async create(params, opts) {
    return this.http.post("/v1/templates", params, opts);
  }
  /**
   * List all templates belonging to the authenticated owner.
   *
   * Auth: Supabase JWT or API key (`rsk_...`)
   *
   * @example
   * ```ts
   * const templates = await client.templates.list();
   * for (const t of templates) {
   *   console.log(t.id, t.name);
   * }
   * ```
   */
  async list(opts) {
    return this.http.get("/v1/templates", opts);
  }
  /**
   * Get a single template by UUID.
   * The template must belong to the authenticated owner.
   *
   * Auth: Supabase JWT or API key (`rsk_...`)
   *
   * @example
   * ```ts
   * const template = await client.templates.get("uuid-here");
   * console.log(template.system);
   * ```
   */
  async get(templateId, opts) {
    return this.http.get(`/v1/templates/${templateId}`, opts);
  }
  /**
   * Update a template. Only provided fields are changed.
   * The template must belong to the authenticated owner.
   *
   * Auth: Supabase JWT or API key (`rsk_...`)
   *
   * @example
   * ```ts
   * const updated = await client.templates.update("uuid-here", {
   *   system: "You are a helpful agent for {{company}}. Be concise.",
   * });
   * ```
   */
  async update(templateId, params, opts) {
    return this.http.patch(`/v1/templates/${templateId}`, params, opts);
  }
  /**
   * Delete a template (hard delete).
   * The template must belong to the authenticated owner.
   *
   * Auth: Supabase JWT or API key (`rsk_...`)
   *
   * @example
   * ```ts
   * await client.templates.delete("uuid-here");
   * ```
   */
  async delete(templateId, opts) {
    return this.http.delete(`/v1/templates/${templateId}`, opts);
  }
};

// src/index.ts
var MeshAPI = class {
  /**
   * Chat completions namespace.
   * Requires a data-plane API key (`rsk_...`).
   *
   * @example
   * ```ts
   * const response = await client.chat.completions.create({
   *   model: "openai/gpt-4o-mini",
   *   messages: [{ role: "user", content: "Hello!" }],
   * });
   * ```
   */
  chat;
  /**
   * Responses namespace — higher-level inference for reasoning models.
   * Requires a data-plane API key (`rsk_...`).
   *
   * Uses `input` (string or message list) instead of `messages`, supports
   * `reasoning.effort` for chain-of-thought, and `max_output_tokens` instead
   * of `max_tokens`. Streaming uses the same SSE chunk format as chat/completions.
   *
   * @example
   * ```ts
   * const response = await client.responses.create({
   *   model: "openai/o4-mini",
   *   input: "Explain the halting problem simply.",
   *   reasoning: { effort: "medium" },
   * });
   * console.log(response.choices[0]?.message.content);
   * ```
   */
  responses;
  embeddings;
  compare;
  files;
  batches;
  /**
   * Models namespace.
   * Accepts either a data-plane API key or a Supabase JWT.
   *
   * @example
   * ```ts
   * const models = await client.models.list();
   * ```
   */
  models;
  /**
   * Templates namespace.
   * Accepts either a data-plane API key or a Supabase JWT.
   *
   * @example
   * ```ts
   * const template = await client.templates.create({ name: "my-template", system: "..." });
   * ```
   */
  templates;
  constructor(config) {
    const http = new HttpClient(config);
    this.chat = new ChatResource(http);
    this.responses = new ResponsesResource(http);
    this.embeddings = new EmbeddingsResource(http);
    this.compare = new CompareResource(http);
    this.files = new FilesResource(http);
    this.batches = new BatchesResource(http);
    this.models = new ModelsResource(http);
    this.templates = new TemplatesResource(http);
  }
};

exports.MeshAPI = MeshAPI;
exports.MeshAPIApiError = MeshAPIApiError;
